cfpb/cfgov-refresh
# Disable image renditions for GIFs

We've experienced problems uploading animated GIFs as Wagtail images. These seem to be caused by problems with the underlying image processing libraries used as part of Wagtail's image renditions logic. This PR bypasses the Wagtail image rendition process for all GIF images, and forces use of the original image file in all template references. 

Normally, anytime the Wagtail [`{% image %}` tag](http://docs.wagtail.io/en/latest/topics/images.html) is used in a template (either Wagtail built-in or our own), some image processing is performed on the image being embedded. This is true even if the "original" format of the image is referred to, i.e. `{% image 'original' %}` (this invokes a file-open-read-write-in-same-format operation). This process can fail on large animated GIF files.

The Wagtail admin makes use of a thumbnail image rendition `{% image max-165x165 %}` in the admin images browser and when selecting an image in a page editor. It also uses [certain predefined](http://docs.wagtail.io/en/latest/topics/images.html#images-embedded-in-rich-text) renditions when including images into a Rich Text block (for example, specifying "full-width" when selecting an image actually sets its width to 800 pixels).

This PR bypasses any actual image processing when GIF files are embedded (technically: image files ending in `.gif`) and instead refers to the original image file while keeping appropriate `width` and `height` parameters in the `<img>` tag. This keeps the appearance of default behavior, although on page views a larger (non-optimized for display size) image is being downloaded.

## Additions

- New `add_images` management command for programmatic adding of images (I created this for testing and thought it might be useful enough to keep around).

## Changes

- Pulled image-related classes out of `v1.models.base` in favor of `v1.models.images` for readability.
- Added mock rendition logic so that GIF files embedded in templates always load the source image.

## Testing

- Create a new local, empty database, and run your local server.
- Navigate to the [image upload page](http://localhost:8000/admin/images/multiple/add/) and upload a large, non-GIF image. After it's been uploaded, navigate back to the [image browse page](http://localhost:8000/admin/images/) and inspect the image element. Note that a thumbnail filter has been applied and that a resized image is being referred to:

![image](https://cloud.githubusercontent.com/assets/654645/20943734/bc1bac04-bbce-11e6-9291-d13ccca678ea.png)

- Next, repeat the same procedure with a large GIF image. Note that the original image is what gets embedded in the page, with appropriately-set `width` and `height` attributes:

![image](https://cloud.githubusercontent.com/assets/654645/20943807/f8085cbc-bbce-11e6-9152-91a2f6e5bd35.png)

- Test the Rich Text block functionality by creating a new BrowsePage and adding a FullWidthText element with a Context block. Within the Context block, select the image icon and select the large PNG file, entering it as "left aligned". Then preview the page and inspect the image element. Note that a `width-500` filter has been applied, and resized image is being embedded in the page:

![image](https://cloud.githubusercontent.com/assets/654645/20944030/f382a5a2-bbcf-11e6-99a6-9490737cd821.png)

- Repeat the previous step, but with a large GIF image. Note that the original image is embedded, but with `width` and `height` attributes set to 500px:

![image](https://cloud.githubusercontent.com/assets/654645/20943982/bb424ba2-bbcf-11e6-98ca-ce1509bce1c8.png)

## Notes

- One discussion-worthy choice that I made here is to explicitly fail page renders if a GIF image is passed to a Wagtail revision function that involves actual image processing beyond just a resize. Examples of these would be `fill` (which resizes an image by padding it to fill a certain size) or `format` which converts an image to a different file format. Another option would be to just pass the original image through (thus ignoring the revision request) and maybe log a warning message; I thought an explicit `RuntimeError` (which would cause a 500 in practice) was a stricter defense of this. @schaferjh @willbarton any thoughts?

## Todos

- Should the site documentation be updated with a mention of this change? If so, where in the documentation tree would be appropriate?

## Checklist
